source/server/listener_manager_impl.cc

```diff
@@ -564,19 +564,29 @@ void ListenerManagerImpl::startWorkers(GuardDog& guard_dog) {
 void ListenerManagerImpl::stopListeners(StopListenersType stop_listeners_type) {
   stop_listeners_type_ = stop_listeners_type;
   for (Network::ListenerConfig& listener : listeners()) {
-    for (const auto& worker : workers_) {
-      if (stop_listeners_type != StopListenersType::InboundOnly ||
-          listener.direction() == envoy::api::v2::core::TrafficDirection::INBOUND) {
-        ENVOY_LOG(debug, "begin stop listener: name={}", listener.name());
-
+    if (stop_listeners_type != StopListenersType::InboundOnly ||
+        listener.direction() == envoy::api::v2::core::TrafficDirection::INBOUND) {
+      std::string listener_name = listener.name();
+      ENVOY_LOG(debug, "begin stop listener: name={}", listener.name());
+      auto workers_pending_stop = std::make_shared<std::atomic<uint32_t>>(workers_.size());
+      for (const auto& worker : workers_) {
         auto existing_warming_listener = getListenerByName(warming_listeners_, listener.name());
         // Destroy a warming listener directly.
         if (existing_warming_listener != warming_listeners_.end()) {
           (*existing_warming_listener)->debugLog("removing warming listener");
           warming_listeners_.erase(existing_warming_listener);
         }
-        worker->stopListener(listener);
-        stats_.listener_stopped_.inc();
+        worker->stopListener(listener, [this, workers_pending_stop, listener_name]() {
```

Comment:
I think maybe you should use the tag here and not the name? The name might be OK but we use tag in all similar post back cases and I think that would be safer? WDYT?